lib/tasks/instrumentation_generator/README.md

```diff
@@ -1,42 +1,63 @@
 # Instrumentation Generator
 
-## Idea
-Create a CLI, similar to Rails’ generators and scaffold scripts, to create the required files and basic classes for instrumenting new libraries. The hope is that the parts of adding new instrumentation that is repetitive can be eliminated and/or reduced so that it takes less time to add new instrumentation and we can spend development time considering the best attributes to collect and interactions to measure. I also hope this project will reduce the overall toil related to adding new instrumentation.
-
-The files we create with little variation besides library and method names are:
-* Lib
-  * Dependency detection
-  * Chain
-  * Prepend
-  * Instrumentation
-* Tests
-  * configuration/newrelic.yml
-  * Envfile
-  * Test file
-
-Furthermore, we also create very similar snippets inside the default source configuration file that could be generated by this project as well.
-
-## Outcome
-A prototype outside the agent has been created that generates the required files to create new instrumentation. This prototype accepts three arguments: name (name of the library), method (method to instrument), args (arguments for the method).
-
-The Ruby gem Thor, a toolkit for building powerful command-line interfaces used in Bundler, Vagrant, Rails and others powers this CLI.
-
 ## Usage
 
 `thor list` can be passed to get a list of all available thor commands
 
 ### instrumentation:scaffold
+
 This task requires one parameter by default: the name of the library or class you are instrumenting. This task generates the basic file structure needed to add new instrumentation to the Ruby agent.
 
+Run it using:
+
+`thor instrumentation:scaffold gem_name`
+
 It accepts two optional parameters:
-* `--method`: the name of a method to instrument. Only one method can be accepted at this time. Defaults to `method_to_instrument`
-* `--args`: the arguments for the instrumented method. A comma-separated list can be given if there are multiple arguments. Defaults to `*args`.
+
+- `--method`: the name of a method to instrument. Only one method can be accepted at this time. Defaults to `method_to_instrument`
+- `--args`: the arguments for the instrumented method. A comma-separated list can be given if there are multiple arguments. Defaults to `*args`.
```

Comment:
I might be misreading it here, but I can't find where the

> Defaults to `*args`.

statement given here is actually fulfilled in code. It seems like the actual default is `nil`, and the generated code ends up leaving of the method's parentheses entirely.

Reply:
I've added a few comments elsewhere in the code to show you how the `--args` option is working. Please let me know if there's anything about the option that seems off!

Reply:
I'm referring to the literal use of `*args` with a `*`, not a `@`.

From reading the documentation, I conclude that if I do not pass `--args`, then the default will be a literal `"*args"` value. So when I look at the template generated methods, I would expect to see this:

```ruby
def generated_method(*args)
  # generated body
end
```

but instead I see this:

```ruby
def generated_method
  # generated body
end
```

when `--args` is left out. I might be understanding it wrong. I just want to make sure that what is stated in the documentation as a default value is actually the real default value (or vice versa, as long as they align).

Reply:
Hmm... could you give me an example of how you're running this? 

When I run `be thor instrumentation:scaffold concurrent-ruby`, I get the following in the `instrumentation.rb` file:
```
    def method_to_instrument_with_tracing(*args)
      # add instrumentation content here
    end
```

Reply:
Ah, I understand why I became confused now.

The `instrumentation.thor` file is indeed applying a default of `*args`, which completely matches the documentation. Everything is great. Sorry for missing that bit. You definitely didn't explain anything badly in comments and all of the code itself is certainly intuitive and great as-is.

When I saw these three lines:

```ruby
        def <%= @method.downcase %><%= "(#{@args})" if @args %>
          <%= @method.downcase %>_with_tracing<%= "(#{@args})" if @args %> do
            <%= @method.downcase %>_without_new_relic<%= "(#{@args})" if @args %>
```

I was picturing what would happen when the `if` checks failed and was incorrectly assuming that there must be a default of `nil` that you are guarding against here. So... is it even possible for the `if` checks to fail given that Thor will always apply a default? Are the checks to protect against a user doing something like `thor instrumentation:add_new_method gem_name --method=method_to_instrument --args=''`?

Reply:
Great point! When I ran that locally, it would still add the parens, so I adjusted the guard statement to be for an empty string rather than a nil check. Take a look at f633c2b5381ec015d93309b6e3323d6de08fa892 for details.